File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/domain/ums/service/EmailVerificationService.kt
package link.yologram.api.v1.domain.ums.service

import link.yologram.api.v1.domain.ums.entity.EmailVerificationCode
import link.yologram.api.v1.domain.ums.exception.EmailVerificationExpiredException
import link.yologram.api.v1.domain.ums.exception.EmailVerificationInvalidException
import link.yologram.api.v1.domain.ums.exception.UserDuplicateException
import link.yologram.api.v1.domain.ums.repository.EmailVerificationCodeRepository
import link.yologram.api.v1.domain.ums.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import kotlin.random.Random

@Service
class EmailVerificationService(
    private val emailVerificationCodeRepository: EmailVerificationCodeRepository,
    private val userRepository: UserRepository,
    private val emailSender: EmailSender,
) {

    @Transactional
    fun sendCode(email: String) {
        if (userRepository.existsByEmail(email)) {
            throw UserDuplicateException()
        }

        emailVerificationCodeRepository.deleteAllByEmail(email)

        val code = generateCode()
        val verification = EmailVerificationCode(
            email = email,
            code = code,
            expiredAt = LocalDateTime.now().plusMinutes(5),
        )
        emailVerificationCodeRepository.save(verification)

        emailSender.sendVerificationCode(email, code)
    }

    @Transactional
    fun verifyCode(email: String, code: String) {
        val verification = emailVerificationCodeRepository.findTopByEmailOrderByCreatedAtDesc(email)
            .orElseThrow { EmailVerificationInvalidException() }

        if (verification.expiredAt.isBefore(LocalDateTime.now())) {
            throw EmailVerificationExpiredException()
        }

        if (verification.code != code) {
            throw EmailVerificationInvalidException()
        }

        verification.verified = true
    }

    private fun generateCode(): String {
        return String.format("%06d", Random.nextInt(1_000_000))
    }
}

File: yologram-api-v2/app/domain/ums/email_verification_service.py
import random
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.core.exception import (
    EmailVerificationExpiredException,
    EmailVerificationInvalidException,
    UserDuplicateException,
)
from app.domain.ums.email_sender import EmailSender
from app.domain.ums.model import EmailVerificationCode
from app.domain.ums.repository import EmailVerificationCodeRepository, UserRepository


class EmailVerificationService:

    def __init__(self, db: Session, email_sender: EmailSender):
        self.repository = EmailVerificationCodeRepository(db)
        self.user_repository = UserRepository(db)
        self.email_sender = email_sender

    def send_code(self, email: str) -> None:
        if self.user_repository.find_by_email(email):
            raise UserDuplicateException()

        self.repository.delete_by_email(email)

        code = f"{random.randint(0, 999999):06d}"
        entity = EmailVerificationCode(
            email=email,
            code=code,
            expired_at=datetime.now() + timedelta(minutes=5),
        )
        self.repository.save(entity)
        self.email_sender.send_verification_code(email, code)

    def verify_code(self, email: str, code: str) -> None:
        entity = self.repository.find_latest_by_email(email)
        if not entity:
            raise EmailVerificationInvalidException()
        if entity.expired_at < datetime.now():
            raise EmailVerificationExpiredException()
        if entity.code != code:
            raise EmailVerificationInvalidException()
        entity.verified = True
